src/shared_realm.cpp

```diff
@@ -921,6 +941,53 @@ template Object Realm::resolve_thread_safe_reference(ThreadSafeReference<Object>
 template List Realm::resolve_thread_safe_reference(ThreadSafeReference<List> reference);
 template Results Realm::resolve_thread_safe_reference(ThreadSafeReference<Results> reference);
 
+#if REALM_ENABLE_SYNC
+bool Realm::init_permission_cache()
+{
+    if (m_permissions_cache) {
+        if (is_in_transaction())
+            m_permissions_cache->clear();
+        return true;
+    }
+    if (m_config.sync_config && m_config.sync_config->is_partial) {
+        m_permissions_cache = std::make_unique<sync::PermissionsCache>(read_group(), m_config.sync_config->user->identity());
+        return true;
+    }
+    return false;
+}
+void Realm::invalidate_permission_cache()
+{
+    if (m_permissions_cache)
+        m_permissions_cache->clear();
+}
+ComputedPrivileges Realm::get_privileges()
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    return static_cast<ComputedPrivileges>(m_permissions_cache->get_realm_privileges());
+}
+ComputedPrivileges Realm::get_privileges(StringData object_type)
+{
+    if (!init_permission_cache())
+        return ComputedPrivileges::All;
+    return static_cast<ComputedPrivileges>(m_permissions_cache->get_realm_privileges() & m_permissions_cache->get_class_privileges(object_type));
```

Comment:
After talking to @simonask we should change this to use `can(Privilegege, obj)` where `obj` is the entry in the `__Class` table for the class in question.

We should use a similar pattern for Realm and Object privileges.